### PR TITLE
Better OOC format + roundend deadchat to Discord

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request

the game doesn't send sufficient chat logs to ooc channel when a round ended. dead chats will be sent as well after a round is finished.
The current format is this:

```
< OOC-Sage channel >
(OOC) AmazingBoy: Where did you die, Boris?
(Dead) Boris Alegro: maints. nobody found me
(Dead) Sings-Over-Birds: I had to hide his body

< OOC-Acacia channel >
(OOC) CuriousQ: low pop acacia is cringe
```

## Why It's Good For The Game

Insufficient information isn't good
also, mixing two server ooc chats into one channel isn't that nice too

## Testing Photographs and Procedure

Seriously this isn't what I can test, but the code might be fine.

## Changelog
:cl:
add: deadchat will be sent to discord as well as long as a round ended.
config: New config line - discord_ooc_channel. used to determine which discord channel OOC messages will be sent to.
server: Sending OOC message to discord now includes dead chats (after round finished only). OOC Discord message will describe whether the message is OOC or deadchat, and which server it was sent from.
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Oh wait it doesn't go to the ooc channel

**Comment by EvilDragonfiend** [comment]
@Crossedfall 
I'd like to ask you to set a config, `#@SERVERABBNAME Bee` to Sage and Acacia to both of the servers - as long as you're fine with this code.

File: code/modules/mob/say.dm [hunk 1]
```diff
@@ -79,6 +79,7 @@
 		return
 	var/spanned = say_quote(message)
 	var/rendered = "<span class='game deadsay'><span class='prefix'>DEAD:</span> <span class='name'>[name]</span>[alt_name] <span class='message'>[emoji_parse(spanned)]</span></span>"
+	send_chat_to_discord("dchat", name, emoji_parse(spanned))
```
Comment (itsmeow) [inline code/modules/mob/say.dm]: send the unparsed version

File: code/modules/admin/verbs/deadsay.dm [hunk 2]
```diff
@@ -26,6 +26,7 @@
 		rank_name = pick(strings(DSAY_NICKNAME_FILE, "ranks", CONFIG_DIRECTORY))
 		admin_name = pick(strings(DSAY_NICKNAME_FILE, "names", CONFIG_DIRECTORY))
 	var/rendered = "<span class='game deadsay'><span class='prefix'>DEAD:</span> <span class='name'>[rank_name]([admin_name])</span> says, <span class='message'>\"[emoji_parse(msg)]\"</span></span>"
+	send_chat_to_discord("dchat", "[rank_name]([admin_name])", emoji_parse(msg))
```
Comment (itsmeow) [inline code/modules/admin/verbs/deadsay.dm]: no parsy

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
While you are messing with OOC can you cache the non-emoji-parsed version and send that? Otherwise you get stuff like this in the #ooc channel.
image

**Comment by EvilDragonfiend** [comment]
thanks good catch. I didn't think that much

**Comment by EvilDragonfiend** [comment]
used ick ock indicator

**Comment by itsmeow** [comment]
> While you are messing with OOC can you cache the non-emoji-parsed version and send that? Otherwise you get stuff like this in the #ooc channel.

You need to edit the original ooc proc as well, it will fix an existing bug

**Comment by itsmeow** [comment]
> added ick ock indicator

I don't like this. Only admins can send ooc during round.

**Comment by EvilDragonfiend** [comment]
> > added ick ock indicator
> 
> I don't like this. Only admins can send ooc during round.

Not really. Admin can turn on ooc chat during a round and it's been always recorded

**Comment by EvilDragonfiend** [comment]
> > While you are messing with OOC can you cache the non-emoji-parsed version and send that? Otherwise you get stuff like this in the #ooc channel.
> 
> You need to edit the original ooc proc as well, it will fix an existing bug

I will check it when I got home

**Comment by EvilDragonfiend** [comment]
Also, I feel like #ooc discord channel can be seperated into two as "ooc-sage" and "ooc-acacia" then we don't need `serverabbname` config but need another `discordoochannel` instead, then the format can be more simple.

**Comment by PowerfulBacon** [comment]
It shouldn't say ICK OCK. If anything you can tell since it says once the round ends

**Comment by EvilDragonfiend** [comment]
> It shouldn't say ICK OCK. If anything you can tell since it says once the round ends

There can be a downstream that turns on ooc chat. That's what I am worrying.

**Comment by EvilDragonfiend** [comment]
Ooc will not be parsed now

File: code/modules/client/verbs/ooc.dm [hunk 3]
```diff
@@ -93,10 +93,21 @@ GLOBAL_VAR_INIT(normal_ooc_colour, "#002eb8")
...
+/proc/send_chat_to_discord(type, sayer, msg)
+	var/discord_channel_name = CONFIG_GET(string/discord_ooc_channel) // check server config file. default value of discord_ooc_channel is `OOC-Bee`
+	if(!discord_channel_name)
+		return
+	switch(type)
+		if("ooc_chat")
+			if(SSticker.current_state == GAME_STATE_FINISHED)
+				discordsendmsg(discord_channel_name, "(OOC) **[sayer]:** [msg]")
+			else // using OOC during an ongoing round..?
+				discordsendmsg(discord_channel_name, "(ICK-OCK) **[sayer]:** [msg]")
```
Comment (itsmeow) [inline code/modules/client/verbs/ooc.dm]: Just send it as regular OOC. Some downstreams allow ooc during round

File: code/modules/client/verbs/ooc.dm [hunk 4]
```diff
@@ -93,10 +93,21 @@ GLOBAL_VAR_INIT(normal_ooc_colour, "#002eb8")
 				else
 					to_chat(C, "[badge_data]<span class='ooc'><span class='prefix'>OOC:</span> <EM>[keyname]:</EM> <span class='message linkify'>[msg]</span></span>")
 	// beestation, send to discord
-	if(holder?.fakekey)
-		discordsendmsg("ooc", "**[holder.fakekey]:** [msg]")
-	else
-		discordsendmsg("ooc", "**[key]:** [msg]")
+	send_chat_to_discord("ooc_chat", holder?.fakekey || key, raw_msg)
```
Comment (itsmeow) [inline code/modules/client/verbs/ooc.dm]: use DEFINES for these strings

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/client/verbs/ooc.dm (hunk 4 above)
Comment (EvilDragonfiend) [reply code/modules/client/verbs/ooc.dm]: do you mean `ooc_chat` thingies?

Comment (itsmeow) [reply code/modules/client/verbs/ooc.dm]: Yes, they're magic strings currently which is bad

Comment (EvilDragonfiend) [reply code/modules/client/verbs/ooc.dm]: done

File: code/modules/admin/verbs/deadsay.dm [hunk 5]
```diff
@@ -26,6 +26,7 @@
 		rank_name = pick(strings(DSAY_NICKNAME_FILE, "ranks", CONFIG_DIRECTORY))
 		admin_name = pick(strings(DSAY_NICKNAME_FILE, "names", CONFIG_DIRECTORY))
 	var/rendered = "<span class='game deadsay'><span class='prefix'>DEAD:</span> <span class='name'>[rank_name]([admin_name])</span> says, <span class='message'>\"[emoji_parse(msg)]\"</span></span>"
+	send_chat_to_discord("dchat", "[rank_name]([admin_name])", msg)
```
Comment (itsmeow) [inline code/modules/admin/verbs/deadsay.dm]: Does not use CHAT_TYPE_DCHAT

File: code/modules/mob/say.dm [hunk 6]
```diff
@@ -79,6 +79,7 @@
 		return
 	var/spanned = say_quote(message)
 	var/rendered = "<span class='game deadsay'><span class='prefix'>DEAD:</span> <span class='name'>[name]</span>[alt_name] <span class='message'>[emoji_parse(spanned)]</span></span>"
+	send_chat_to_discord("dchat", name, spanned)
```
Comment (itsmeow) [inline code/modules/mob/say.dm]: same here

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
well, auxmos dependency is pretty fucked right now. can't fix the file for now.

**Comment by EvilDragonfiend** [comment]
done

File: config/config.txt [hunk 7]
```diff
@@ -618,3 +618,6 @@ ELASTICSEARCH_METRICS_APIKEY thisIsSomethingThatsBased64Encoded==
 
 ## Enable/disable the mentor mrat verb ("Request Mentor Assistance")
 ENABLE_MRAT
+
+## Discord OOC channel name. As long as this is enabled, the game will try to send OOC messages to the discord channel. change 'OOC-Bee' into something else
+#DISCORD_OOC_CHANNEL OOC-Bee
```
Comment (PowerfulBacon) [inline config/config.txt]:
```suggestion
#DISCORD_OOC_CHANNEL ooc
```

**Review by PowerfulBacon — DISMISSED** [review]

**Comment by EvilDragonfiend** [comment]
Change done.

**Review by PowerfulBacon — DISMISSED** [review]

**Comment by PowerfulBacon** [comment]
needs cfg applied

**Comment by EvilDragonfiend** [comment]
it'll be lovely if we can have two separated channels for sage and acacia

**Review by itsmeow — DISMISSED** [review]

**Comment by itsmeow** [comment]
@Crossedfall

**Review by Crossedfall — CHANGES_REQUESTED** [review]
You can't configure the Discord channel name through DM. It's a simple webhook, the channel is determined by the webhook URL and cannot be configured externally.

You can see how the Discord relay works here: https://github.com/BeeStation/BeeStation-Hornet/tree/master/tools/discord

**Comment by EvilDragonfiend** [comment]
ah, okay, then I will change messages to have configured names like
`[SG] (OOC) AmazingBoy: Where did you die, Boris?`
`[AC] (OOC) boiboi123: low pop acacia`

**Review by itsmeow — APPROVED** [review]
Seems a good enough solution.

**Review by PowerfulBacon — APPROVED** [review]
Needs config update applied

**Review by Crossedfall — APPROVED** [review]
Looked good during TM.

**Comment by Crossedfall** [comment]
Ack. I forgot to test the round-end deadchat part before merging. That unfortunately was not perfect.

```
[AC] (Dead) Rachiplusia Osseogrisea: says: "<span class=' '>test test test</span>"
```

Not terrible though and should be easy to patch.

**Pull request merged**